sequence_task_1/student/include/impl.h

```diff
@@ -19,32 +20,121 @@ template<class Container>
 class StaticPlaylist
 {
 public:
-    /** @todo Member traits */
+    /** Member traits */
+    /*
+     // simple ...
+    typedef Song         value_type;
+    typedef Song&        reference;
+    typedef const Song&  const_reference;
+    typedef size_t	     size_type;
+    typedef ptrdiff_t    difference_type;
+    */
+    typedef typename Container::value_type       value_type;
+    typedef typename Container::reference        reference;
+    typedef typename Container::const_reference  const_reference;
+    typedef typename Container::size_type        size_type;
+    typedef typename Container::difference_type  difference_type;
+    //i don't know any other way to iterators
+    typedef typename Container::iterator        iterator;
+    typedef typename Container::const_iterator  const_iterator;
 
-    /** @todo Iterators */
+    /** Iterators */
+    const_iterator begin() const
+    {
+        return m_tracklist.begin();
+    }
 
-    /** @todo Constructor from any reversible sequence container */
+    const_iterator end() const
+    {
+        return m_tracklist.end();
+    }
 
-    /** @todo Assignment from any reversible sequence container */
+    StaticPlaylist()
+    : m_tracklist()
+    {
+
+    }
+
+    /** Constructor from any reversible sequence container */
+    //any container? Cheking needed AnyContainer is supported in compile time, like static_assert. Out of scope.
+    template<class AnyContainer>
+    StaticPlaylist(const AnyContainer& ac)
+    : m_tracklist(ac.begin(), ac.end())
+    {
 
-    /** @todo Add track from initializer */
+    }
+
+    StaticPlaylist( const StaticPlaylist& other )
+    : m_tracklist(other.m_tracklist.begin(), other.m_tracklist.end())
+    {
+
+    }
+
+    /** Assignment from any reversible sequence container */
+    //any container? Cheking needed AnyContainer is supported in compile time, like static_assert. Out of scope.
+    template<class AnyContainer>
+    StaticPlaylist& operator=(const AnyContainer& ac)
+    {
+        m_tracklist.clear();
+        m_tracklist.insert(m_tracklist.begin(), ac.begin(), ac.end());
+        return *this;
+    }
+
+    StaticPlaylist& operator=( const StaticPlaylist& other )
+    {
+        if (this != &other && m_tracklist != other.m_tracklist)
+        {
+            m_tracklist = other.m_tracklist;
+        }
+        return *this;
+    }
+
+    /** Add track from initializer */
     template<class... Args>
-    const Song& play(Args&&... songData);
+    const Song& play(Args&&... songData)
+    {
+        m_tracklist.emplace_back(songData...);
```

Comment:
Use std::forward
https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#f19-for-forward-parameters-pass-by-tp-and-only-stdforward-the-parameter

Reply:
accept